File: SphDemo/Partical.py
import math
import kenelFcunction
from typing import List



#1. compute location of every particals
#2. compute density and pressrue of every particals
#3. compute pressure, Viscous coefficient of every particals
#4. compute velocity of every particals according to the total force of every particals
#5. continue step 1 

__c = 200
__initDensity : float = 100
__viscosity : float = 100
__gravityY : float = -9.8



class Partical(object):
  posX = 0
  posY = 0
  velX = 0
  velY = 0
  accX = 0
  accY = 0
  mass = 100
  density = __initDensity


  pressure = 0
  tauXX, tauXY, tauYY, tauYX = 0


  def __init__(self, x, y):
    self.posX = x
    self.posY = y



  def computeDensity(self, neigborList: list['Partical']) -> float:
    dpdt = 0

    for point in neigborList:
      dwdx, dwdy = kenelFcunction.kenelDerivative(self, point)
      dpdt += self.density * point.mass / point.density \
       * ((self.velX - point.velX) * dwdx + (self.velY - point.velY) * dwdy)

    return dpdt



  def computePressrue(self):
    B = 32
    self.pressure = B * ((self.density / __initDensity) ** 7 - 1)
    #self.pressure = self.density * __c **2


  def computeViscous(self, neigborList: list['Partical']):
    dvdx = 0
    dvdy = 0
    for point in neigborList:
      dwdx, dwdy = kenelFcunction.kenelDerivative(self, point)
      dvdx += point.mass / point.density * point.velX * dwdx
      dvdy += point.mass / point.density * point.vely * dwdy

    self.tauXX = __viscosity * (2 * dvdx - 2/3 * (dvdx + dvdy))
    self.tauYY = __viscosity * (2 * dvdy - 2/3 * (dvdx + dvdy))
    self.tauXY = __viscosity * (dvdx + dvdy)
    self.tauYX = __viscosity * (dvdx + dvdy)



  def computeVelocity(self, neigborList: list['Partical']):
    dvxdt = 0;
    dvydt = 0;

    for point in neigborList:
      dwdx, dwdy = kenelFcunction.kenelDerivative(self, point)
      temp = point.density / (self.density * point.density)

      dvxdt += - temp * (self.pressure + point.pressure) * dwdx
      dvxdt += temp * (self.tauXX + point.tauXX) * dwdx
      dvxdt += temp * (self.tauXY + point.tauXY) * dwdy

      dvydt += - temp * (self.pressure + point.pressure) * dwdy
      dvydt += temp * (self.tauYY + point.tauYY) * dwdy
      dvydt += temp * (self.tauYX + point.tauYX) * dwdx

    return dvxdt, dvydt
  


  #the followings are deprecated
  def computePressureGradient(self, neigborList: list['Partical']) -> float:
    #the important is to confine the direction of the pressure
    #this use the pressure gradient equation from the papar of MPS
    #and plus the power of mass 
    gradientX = 0
    gradientY = 0

    for point in neigborList:
      w = self.kenelFunction(point)
      distance = math.sqrt( (self.posX - point.posX) **2 + (self.posY - point.posY) **2 )
      temp = point.mass / point.density * w * (self.pressure - point.pressure) / distance
      gradientX += temp * (self.posX - point.posX)
      gradientY += temp * (self.posY - point.posY)

    return gradientX, gradientY



  def computeViscousForceGradient(self, neigborList: list['Partical']):
    gradientX = 0
    gradientY = 0

    for point in neigborList:
      d2wdx2, d2wdy2, d2wdxdy = kenelFcunction.kenelDerivative2Order(self, point)

      #inclue the normal stress and the shear stress
      vfX = __viscosity * point.mass / point.density \
        * (point.velX * d2wdxdy + point.velY * d2wdy2 \
        + 4/3 * point.velX * d2wdx2 - 2/3 * point.velY * d2wdxdy)
      vfY = __viscosity * point.mass / point.density \
        * (point.velY * d2wdxdy + point.velX * d2wdx2 \
        + 4/3 * point.velY * d2wdy2 - 2/3 * point.velX * d2wdxdy)
      gradientX += vfX
      gradientY += vfY

    return gradientX, gradientY



  def computeTotalForceGradient(self, neigborList: list['Partical']):
    pressureGradientX, pressureGradientY = self.computePressureGradient(neigborList)
    viscousForceGradientX, viscousForceGradientY = self.computeViscousForceGradient(neigborList)

    self.beForceGradientX = -pressureGradientX + viscousForceGradientX
    self.beForceGradientY = -pressureGradientY + viscousForceGradientY

File: SphInCsharp/Partical.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;



namespace SphInCsharp {
  internal class Partical {
    public static double _h = 10.0f;
    static double _ad = 15.0f / (7 * Math.PI * _h * _h);

    public static double _c = 10f;
    public static double _initDensity = 1000.0f;
    public static double _viscosity = 1e10; // 0.001f
    public static double _gravityY = -9.8f;

    public double posX = 0.0f;
    public double posY = 0.0f;
    public double velX = 0.0f;
    public double velY = 0.0f;
    public double mass = 100.0f;
    public double density = _initDensity;

    double pressure = 100f;
    double tauXX, tauXY, tauYY, tauYX = 0.0f;


    public Partical(double x, double y, double mass) {
      posX = x;
      posY = y;
      this.mass = mass;
    }


    double kenel(in Partical other) {
      double xdiff = this.posX - other.posX;
      double ydiff = this.posY - other.posY;
      double dis = Math.Sqrt(xdiff * xdiff + ydiff * ydiff) / _h;

      double w = 0.0f;
      if (dis < 1.0f)
        w = _ad * (2 / 3 - dis * dis + 0.5 * dis * dis * dis);
      else if (dis < 2.0f)
        w = _ad * 1 / 6 * Math.Pow((2 - dis), 3);
      return w;
    }


    Tuple<double, double> kenelDerivative(in Partical other) {
      double xdiff = this.posX - other.posX;
      double ydiff = this.posY - other.posY;
      double rh = Math.Sqrt(xdiff * xdiff + ydiff * ydiff);
      double dis = rh / _h;

      if (rh == 0.0) 
        return new Tuple<double, double>(0, 0);

      if(dis > 2)
        return new Tuple<double, double>(0, 0);

      double drdx = 2 * (this.posX - other.posX) / (_h * rh);  //partial derivative of r by x
      double drdy = 2 * (this.posY - other.posY) / (_h * rh);  //partial derivative of r by y
      double dwdr = 0.0;
      if (dis < 1)
        dwdr = _ad * (3 / 2 * dis * dis - 2 * dis);   //partial derivative of w by r
      if (dis < 2)
        dwdr = -_ad * 1 / 2 * (2 - dis) * (2 - dis);

      double dwdx = dwdr * drdx;
      double dwdy = dwdr * drdy;
      return new Tuple<double, double>(dwdx, dwdy);
    }


    public double computeDensity(in List<Partical> neigborList) {
      double dpdt = 0;

      foreach (var point in neigborList) {
        Tuple<double, double> ddd = kenelDerivative(point);
        double dwdx = ddd.Item1;
        double dwdy = ddd.Item2;
        dpdt += this.density * point.mass / point.density
         * ((this.velX - point.velX) * dwdx + (this.velY - point.velY) * dwdy);
      }

      return dpdt;
    }


    public void computePressrue() {
      double B = 32;
      //this.pressure = B * (Math.Pow(this.density / _initDensity, 7) - 1);
      this.pressure = this.density * _c * _c;
    }


    public void computeViscous(in List<Partical> neigborList) {
      double dvdx = 0;
      double dvdy = 0;
      foreach(var point in neigborList) {
        Tuple<double, double> ddd = kenelDerivative(point);
        dvdx += point.mass / point.density * point.velX * ddd.Item1;
        dvdy += point.mass / point.density * point.velY * ddd.Item2;
      }


      //which one is currect?
      this.tauXX = _viscosity * (-2 / 3 * (dvdx + dvdy));
      this.tauYY = _viscosity * (-2 / 3 * (dvdx + dvdy));
      //this.tauXX = _viscosity * (2 * dvdx - 2 / 3 * (dvdx + dvdy));
      //this.tauYY = _viscosity * (2 * dvdy - 2 / 3 * (dvdx + dvdy));

      this.tauXY = _viscosity * (dvdx + dvdy);
      this.tauYX = _viscosity * (dvdx + dvdy);
    }


    public Tuple<double, double> computeVelocity(in List<Partical> neigborList) {
      double dvxdt = 0;
      double dvydt = 0;

      foreach(var point in neigborList) {
        Tuple<double, double> ddd = kenelDerivative(point);
        double dwdx = ddd.Item1;
        double dwdy = ddd.Item2;
        double temp = point.density / (this.density * point.density);

        dvxdt += -temp * (this.pressure + point.pressure) * dwdx;
        dvxdt += temp * (this.tauXX + point.tauXX) * dwdx;
        dvxdt += temp * (this.tauXY + point.tauXY) * dwdy;

        dvydt += -temp * (this.pressure + point.pressure) * dwdy;
        dvydt += temp * (this.tauYY + point.tauYY) * dwdy;
        dvydt += temp * (this.tauYX + point.tauYX) * dwdx;
      }

      return new Tuple<double, double>(dvxdt, dvydt);
    }
  }
}
